packages/service-library/src/data-providers/on-demand-page-scan-run-result-provider.ts

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { inject, injectable } from 'inversify';
+
+import { GuidUtils, HashGenerator } from 'common';
+
+import { CosmosContainerClient, cosmosContainerClientTypes } from 'azure-services';
+import { flatMap, groupBy } from 'lodash';
+import { ItemType, OnDemandPageScanResult } from 'storage-documents';
+
+@injectable()
+export class OnDemandPageScanRunResultProvider {
+    public static readonly partitionKeyPreFix = 'pageScanRunResult';
+
+    constructor(
+        @inject(HashGenerator) private readonly hashGenerator: HashGenerator,
+        @inject(GuidUtils) private readonly guidUtils: GuidUtils,
+        @inject(cosmosContainerClientTypes.OnDemandScanRunsCosmosContainerClient)
+        private readonly cosmosContainerClient: CosmosContainerClient,
+    ) {}
+
+    public async createScanRuns(scanRuns: OnDemandPageScanResult[]): Promise<void> {
+        const scanRunsByPartition = groupBy(scanRuns, scanRun => {
+            this.setSystemProperties(scanRun);
+
+            return scanRun.partitionKey;
+        });
+
+        await Promise.all(
+            Object.keys(scanRunsByPartition).map(async pKey => {
+                return this.cosmosContainerClient.writeDocuments(scanRunsByPartition[pKey], pKey);
+            }),
+        );
+    }
+
+    public async readScanRuns(scanIds: string[]): Promise<OnDemandPageScanResult[]> {
+        const scanIdsByPartition = groupBy(scanIds, scanId => {
+            return this.getPartitionKey(scanId);
+        });
+
+        const response = await Promise.all(
+            Object.keys(scanIdsByPartition).map(async pKey => {
+                return this.cosmosContainerClient.executeQueryWithContinuationToken<OnDemandPageScanResult>(async token => {
+                    return this.cosmosContainerClient.queryDocuments<OnDemandPageScanResult>(
+                        this.getReadScanQueryForScanIds(scanIdsByPartition[pKey]),
+                        token,
+                        pKey,
+                    );
+                });
+            }),
+        );
+
+        return flatMap(response);
+    }
+
+    public async updateScanRun(pageScanResult: OnDemandPageScanResult): Promise<void> {
+        this.setSystemProperties(pageScanResult);
+
+        await this.cosmosContainerClient.mergeOrWriteDocument(pageScanResult);
+    }
+
+    private getReadScanQueryForScanIds(scanIds: string[]): string {
+        const orConditions = scanIds.map(scanId => {
+            return `c.id = '${scanId}'`;
+        });
+
+        return `select * from c where ${orConditions.join(' or ')}`;
```

Comment:
The query has limits, if we change batch size this can be affected, we may do simple limit check here https://docs.microsoft.com/en-us/azure/cosmos-db/concepts-limits#sql-query-limits